pkg/collector/corechecks/system/cpu/cpu_windows.go

```diff
@@ -77,24 +78,65 @@ type processorPDHCounter struct {
 	// see https://github.com/DataDog/datadog-agent/pull/8881
-	// On machines with more than 1 NUMA node, it uses "Processor Information",
-	// otherwise it uses "Processor" (e.g. in containers).
-	// Note we use "processor information" instead of "processor" because on multi-processor machines the later only gives
-	// you visibility about other applications running on the same processor as you
-	err := counter.PdhEnglishSingleInstanceCounter.AddToQuery(query)
+	// Unfortunately, since the counterset exists checking the PdhAddEnglishCounter result
+	// alone is insufficient. We must perform additional checks to ensure the counterset
+	// is functioning properly.
+
+	// Add counter to the Query
+	var origObjectName = counter.ObjectName
+	counter.ObjectName = objectName
+	err = counter.PdhEnglishSingleInstanceCounter.AddToQuery(query)
 	if err != nil {
-		counter.ObjectName = "Processor"
-		err = counter.PdhEnglishSingleInstanceCounter.AddToQuery(query)
+		// PdhAddEnglishCounter failed, restore original object name and return
+		counter.ObjectName = origObjectName
+		return err
+	}
+	defer func() {
 		if err != nil {
-			// Processor failed, too. Restore default objectName so we can try it again next time.
-			counter.ObjectName = "Processor Information"
-
+			// failed, restore original object name
+			counter.ObjectName = origObjectName
+			// Remove the counter from the query
+			tmpErr := counter.Remove()
+			if tmpErr != nil {
+				log.Warnf("cpu.Check: Failed to remove counter \\%s(%s)\\%s from query. %v", objectName, counter.InstanceName, counter.CounterName, tmpErr)
+			}
+		}
+	}()
+
+	// Add succeeded, check if the counter is working (see above comment)
+	// Must call PdhCollectQueryData() twice before GetValue() will succeed.
+	// This call may fail for the first counter added to the query. Once at least one Processor
+	// counter is successfully added this call will succeed and we must depend on the GetValue() call.
+	for i := 0; i < 2; i++ {
+		err = pdhutil.PdhCollectQueryData(query.Handle)
+		if err != nil {
+			return err
 		}
 	}
+	_, err = counter.GetValue()
+	return err
+}
+
+func (counter *processorPDHCounter) AddToQuery(query *pdhutil.PdhQuery) error {
+	// Configure the PDH counter according to the running environment.
+	// This check defaults to using the "Processor Information" counterset when it is available,
+	// as it is the newer version of the "Processor" counterset and has support for more features.
+	// https://techcommunity.microsoft.com/t5/running-sap-applications-on-the/windows-2008-r2-performance-monitor-8211-processor-information/ba-p/367007
+	// See addToQueryWithExtraChecks for more details.
+	err := counter.addToQueryWithExtraChecks(query, "Processor Information")
+	if err != nil {
+		log.Warnf("cpu.Check: Error initializing counter \\%s(%s)\\%s: %v. This error is expected in Windows containers. Trying Processor counterset as a fallback.", counter.ObjectName, counter.InstanceName, counter.CounterName, err)
```

Comment:
Just curious if \\%s(%s)\\%s: is worth to be expressed as counter's method path to be used here and few other places (and it could be polymorphic for single/multiple).

Reply:
going to hold this change until later because I think it requires too many changes to the structs/interfaces.